lib/Dialect/FIRRTL/Transforms/AssignOutputDirs.cpp

```diff
@@ -0,0 +1,158 @@
+  if (moduleOutputDir.starts_with(outputDir))
+    moduleOutputDir.erase(moduleOutputDir.begin(),
+                          moduleOutputDir.begin() + outputDir.size());
+}
+
+static void makeCommonPrefix(SmallString<64> &a, StringRef b) {
+  // truncate 'a' to the common prefix of 'a' and 'b'.
+  size_t i = 0;
+  size_t e = std::min(a.size(), b.size());
+  for (; i < e; ++i)
+    if (a[i] != b[i])
+      break;
+  a.resize(i);
+
+  // truncate 'a' so it ends on a directory seperator.
+  auto sep = path::get_separator();
+  while (!a.empty() && !a.ends_with(sep))
+    a.pop_back();
+}
+
+static void makeCommonPrefix(StringRef outputDir, SmallString<64> &a,
+                             OutputFileAttr attr) {
+  if (attr) {
+    SmallString<64> b(attr.getFilename());
+    makeAbsolute(outputDir, b);
+    makeCommonPrefix(a, b);
+  } else {
+    makeCommonPrefix(a, outputDir);
+  }
+}
+
+static OutputFileAttr getOutputFile(Operation *op) {
+  return op->getAttrOfType<hw::OutputFileAttr>("output_file");
+}
+
+namespace {
+struct AssignOutputDirsPass
+    : public AssignOutputDirsBase<AssignOutputDirsPass> {
+  AssignOutputDirsPass() = default;
+  AssignOutputDirsPass(StringRef outputDir) : AssignOutputDirsPass() {
+    outputDirOption = std::string(outputDir);
+  }
+
+  void runOnOperation() override;
+};
+} // namespace
+
+void AssignOutputDirsPass::runOnOperation() {
+  SmallString<64> outputDir(outputDirOption);
+  if (fs::make_absolute(outputDir)) {
+    llvm::errs()
+        << "error: failed to convert the outputDir to an absolute path\n";
+    signalPassFailure();
+    return;
+  }
+  path::remove_dots(outputDir, true);
+  auto sep = path::get_separator();
+  if (!outputDir.ends_with(sep))
+    outputDir.append(sep);
+
+  bool changed = false;
+
+  DenseSet<InstanceGraphNode *> visited;
+  for (auto *root : getAnalysis<InstanceGraph>()) {
+    for (auto *node : llvm::inverse_post_order_ext(root, visited)) {
+      auto module = dyn_cast<FModuleOp>(node->getModule().getOperation());
+      if (!module || module->getAttrOfType<hw::OutputFileAttr>("output_file") ||
+          module.isPublic())
+        continue;
+
+      // Get the output directory of the first parent, and then fold the current
+      // output directory with the LCA of all other discovered output
+      // directories.
+      SmallString<64> moduleOutputDir;
+      auto i = node->usesBegin();
+      auto e = node->usesEnd();
+      for (; i != e; ++i) {
+        if (auto parent = dyn_cast<FModuleOp>((*i)->getParent()->getModule())) {
+          auto file = getOutputFile(parent);
+          if (file) {
+            moduleOutputDir = file.getDirectory();
+            makeAbsolute(outputDir, moduleOutputDir);
+          } else {
+            moduleOutputDir = outputDir;
+          }
+          ++i;
+          break;
+        }
+      }
+      for (; i != e; ++i) {
+        if (auto parent =
+                dyn_cast<FModuleOp>((*i)->getParent()->getModule<FModuleOp>()))
```

Comment:
Probably the dyn_cast AND the `getModule<FModuleOp>` are redundant? Whichever style works for me, but use same above too please :+1:.